File: DesignPattenStudy/DecoratorPattern/Java/Concrete/IceDecorator.java
package DecoratorPattern.Java.Concrete;

import DecoratorPattern.Java.Abstract.Component;
import DecoratorPattern.Java.Abstract.Decorator;

public class IceDecorator extends Decorator {

    // coffee ice blended
    public IceDecorator(Component component) {

        super(component);
    }

    @Override
    public Integer addPrice() {

        return super.addPrice() + new EspressoDecorator(new BaseComponent()).addPrice() + 1000;
    }

    @Override
    public String showItem() {
        return super.showItem() + " + Ice : CoffeeIceBlended";
    }
}

File: DesignPattenStudy/DecoratorPattern/Kotlin/Concrete/IceDecorator.kt
package DecoratorPattern.Kotlin.Concrete

import DecoratorPattern.Kotlin.Astract.Component
import DecoratorPattern.Kotlin.Astract.Decorator

class IceDecorator(override val component: Component) : Decorator(component) {

    override fun addPrice(): Int = super.addPrice() + EspressoDecorator(BaseComponent()).addPrice() + 1000

    override fun showItem(): String = "${super.showItem()} + Ice : CoffeeIceBlended"
}